### PR TITLE
Fix ICE on const eval of union field

### LINKED ISSUE #47788: ICE when combining const (fn) and unions
This code posted by @sfleischman105 in #24111 currently compiles:

```rust
#![feature(const_fn)]

use std::time::Instant;

pub struct GlobalTimer {
    time: Instant,
}

union DummyUnion {
    time: Instant,
    junk: u32
}

impl GlobalTimer {
    pub const fn init() -> GlobalTimer {
        const DUMMY: DummyUnion = DummyUnion {
            junk: 0
        };
        
        const UNINIT: Instant = unsafe {
            DUMMY.time
        };
        
        // Ta-da!
        GlobalTimer {
            time: UNINIT,
        }
    }
}

fn main() {
    //GlobalTimer::init();
}
```

With this suspicious (and wrong?) warning:
```
warning: constant evaluation error: nonexistent struct field
  --> src/main.rs:20:33
   |
20 |           const UNINIT: Instant = unsafe {
   |  _________________________________^
21 | |             DUMMY.time
22 | |         };
   | |_________^
   |
   = note: #[warn(const_err)] on by default
```

However, trying to use the `const fn` by uncommenting the line in `main` causes an ICE:
```
error: internal compiler error: librustc_trans/type_of.rs:386: TyLayout::llvm_field_index(TyLayout { ty: DummyUnion, details: LayoutDetails { variants: Single { index: 0 }, fields: Union(2), abi: Aggregate { sized: true }, align: Align { abi: 3, pref: 3 }, size: Size { raw: 16 } } }): not applicable
```

### PR DESCRIPTION
**Pull request by etaoins**
MIR's `Const::get_field()` attempts to retrieve the value for a given field in a constant. In the case of a union constant it was falling through to a generic `const_get_elt` based on the field index. As union fields don't have an index this caused an ICE in `llvm_field_index`.

Fix by simply returning the current value when accessing any field in a union. This works because all union fields start at byte offset 0.

The added test uses `const_fn` it ensure the field is extracted using MIR's const evaluation. The crash is reproducible without it, however.

Fixes #47788

r? @eddyb

### CODE REVIEW

File: src/test/run-pass/union/union-const-eval-field.rs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![feature(const_fn)]
+
+union DummyUnion {
+    field1: i32,
+    field2: i32,
```
Comment (eddyb) [inline src/test/run-pass/union/union-const-eval-field.rs]: What happens if you make them different types, say, `i64` for the second field`, and maybe do some arithmetic on it?

**Comment by eddyb** [comment]
Thank you for the PR, however #46882 should fix the ICE and implement the correct semantics in *all* cases, so I'm not sure we should merge this partial fix. cc @oli-obk

**Comment by etaoins** [comment]
@eddyb 

I can confirm that #46882 fixes this. My only concern is this a regression against stable as it's possible to crash nightly and beta (but not 1.23) with the following:

```rust
union DummyUnion {
    field1: i32,
    field2: i32,
}

const UNION: DummyUnion = DummyUnion { field1: 1 };

fn read_field1() -> i32 {
    const FIELD1: i32 = unsafe { UNION.field1 };
    FIELD1
}

fn main() {
    read_field1();
}
```

**Comment by eddyb** [comment]
Yes but what does stable do if the types aren't the same? It can't possibly work. Oh well.

cc @rust-lang/compiler

**Comment by eddyb** [comment]
Also I think your test is still subtly cheating, try an union between `i64` and `(i32, i32)`, where you initialize the former and read the latter. Should fail on both stable and this PR.

**Comment by etaoins** [comment]
This passes on both stable and this PR but crashes on nightly and beta:

```rust
type Field1 = u64;
type Field2 = (u32, u32);

union DummyUnion {
    field1: Field1,
    field2: Field2,
}

const UNION: DummyUnion = DummyUnion {
    field1: 0xdeadbeeff00fb00f,
};

fn read_field1() -> Field1 {
    const FIELD1: Field1 = unsafe { UNION.field1 };
    FIELD1
}

fn read_field2() -> Field2 {
    const FIELD2: Field2 = unsafe { UNION.field2 };
    FIELD2
}

fn main() {
    assert_eq!(read_field1(), 0xdeadbeeff00fb00f);

    let (first, second) = read_field2();
    assert_eq!(first, 0xf00fb00f);
    assert_eq!(second, 0xdeadbeef);
}
```

**Comment by petrochenkov** [comment]
This is not supposed to work on stable.
There's a test ensuring this doesn't work - https://github.com/rust-lang/rust/blob/master/src/test/ui/union/union-const-eval.rs.
Const functions however change the intended error into a warning somehow and it's a bug.

**Comment by eddyb** [comment]
@etaoins Sorry, what I meant was this:
```rust
const FIELD2_1: u32 = unsafe { UNION.field2.1 };
```

**Comment by oli-obk** [comment]
This code will get removed by #46882 and the underlying issue will be fixed. I'll add this test to my PR, it should simply work

**Comment by etaoins** [comment]
@eddyb

Yup, that finally kills it. Running this will segfault on stable and fail the assertion on my PR:

```rust
type Field1 = u64;
type Field2 = (u32, u32);

union DummyUnion {
    field1: Field1,
    field2: Field2,
}

const UNION: DummyUnion = DummyUnion {
    field1: 0xdeadbeeff00fb00f,
};

fn read_field2_1() -> u32 {
    const FIELD2_1: u32 = unsafe { UNION.field2.1 };
    FIELD2_1
}

fn main() {
    assert_eq!(read_field2_1(), 0xdeadbeef);
}
```

**Comment by eddyb** [comment]
I think we should merge this and backport it, it shouldn't allow anything that didn't work before.

**Comment by etaoins** [comment]
@eddyb What's the process for backporting? Should I open a PR against the beta branch?

**Comment by eddyb** [comment]
@etaoins We first have to accept it and  then we typically batch them.

**Comment by nikomatsakis** [comment]
Accepting for backport.

**Pull request merged**